kratos/modeler/internals/skin_intersection.h

```diff
@@ -0,0 +1,289 @@
+    {
+        using PointType = typename GeometryType::PointType;
+        std::vector<GeometryPointerType> cut, trimmed_geometries;
+        Vector nodal_distances;
+
+        for (auto iter = rGeometries.begin(); iter < rGeometries.end();  ++iter) {
+            auto& r_geom = **iter;
+            cut.resize(0);
+
+            std::size_t points_under = SignedDistanceUnder(nodal_distances, r_geom, rLowPoint, Component);
+            if (points_under == 3) {
+                trimmed_geometries.push_back(*iter);
+            }
+            else if (points_under > 0) {
+                auto splitter = DivideTriangle3D3<PointType>(r_geom, nodal_distances);
+                splitter.GenerateDivision();
+                cut = splitter.GetPositiveSubdivisions();
+                trimmed_geometries.insert(trimmed_geometries.end(), cut.begin(), cut.end());
+            }
+        }
+
+        return trimmed_geometries;
+    }
+
+    template<class GeometryType, class GeometryPointerType = typename GeometryType::Pointer>
+    static std::vector<GeometryPointerType> TrimLower(
+        const std::vector<GeometryPointerType>& rGeometries,
+        const Point& rHighPoint,
+        std::size_t Component)
+    {
+        using PointType = typename GeometryType::PointType;
+        std::vector<GeometryPointerType> cut, trimmed_geometries;
+        Vector nodal_distances;
+
+        for (auto iter = rGeometries.begin(); iter < rGeometries.end();  ++iter) {
+            auto& r_geom = **iter;
+            cut.resize(0);
+
+            std::size_t points_over = SignedDistanceOver(nodal_distances, r_geom, rHighPoint, Component);
+            if (points_over == 3) {
+                trimmed_geometries.push_back(*iter);
+            }
+            else if (points_over > 0) {
+                auto splitter = DivideTriangle3D3<PointType>(r_geom, nodal_distances);
+                splitter.GenerateDivision();
+                cut = splitter.GetPositiveSubdivisions();
+                trimmed_geometries.insert(trimmed_geometries.end(), cut.begin(), cut.end());
+            }
+        }
+
+        return trimmed_geometries;
+    }
+
+    // Each component of rIntegrationPointData is a 4-component vector
+    // Components 0-2 are the GLOBAL coordinates of the integration point
+    // Component 3 is the integration point weight in terms of area (DetJ*Gauss_W)
+    void AddTriangleIntegrationPointData(
+        std::vector<Vector>& rIntegrationPointData,
+        const SplitGeometryType& rGeom,
+        const GeometryData::IntegrationMethod Method) const
+    {
+        const std::size_t num_nodes = rGeom.PointsNumber();
+        const auto& r_integration_points = rGeom.IntegrationPoints(Method);
+        const std::size_t num_gauss = r_integration_points.size();
+
+        Matrix shape_functions = rGeom.ShapeFunctionsValues(Method);
+        Vector det_j;
+        rGeom.DeterminantOfJacobian(det_j, Method);
+
+        std::size_t offset = rIntegrationPointData.size();
```

Comment:
This is not used until the 2nd loop, move it with the other variables at the top or just before the 2nd loop.